### PR TITLE
Refactor JSON parser to run a list of transformations

### PR DESCRIPTION
**Pull request by DixonDs**
### Description
I noticed a bunch of attempted fixes to JSON parsing, resulting in messy logic that is hard to follow. This is my attempt to refactor it by introducing a uniform list of JSON transformations which we run in a loop until we get a valid JSON. This should make it easier to add/remove other JSON fixes.

I tried to keep the existing logic as close as possible as the first step. If this PR is accepted, the next step would be to consolidate all JSON parsing logic in `jsonparser.py`, meaning to merge `json_utils.py` and extract other JSON parsing logic from `main.py`.

### Testing
Executed existing tests and made a short run of AutoGPT. I haven't added new tests since this should not change the existing functionality.

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

**Review by WolfXae — APPROVED** [review]

**Review by nponeccop — DISMISSED** [review]

**Comment by nponeccop** [comment]
@DixonDs There are conflicts

**Comment by DixonDs** [comment]
@nponeccop Fixed

**Review by nponeccop — APPROVED** [review]

**Review by nponeccop — APPROVED** [review]

**Comment by nponeccop** [comment]
@DixonDs There are conflicts now

**Comment by DixonDs** [comment]
@nponeccop  Fixed. It will be probably a never-ending loop with all the PRs accepted adding a space here and an empty line there :)

**Comment by nponeccop** [comment]
It won't. You are in Batch 4. Batch 4 will be merged tomorrow if you are available to fix conflicts.

**Review by nponeccop — DISMISSED** [review]

**Comment by richbeales** [comment]
ERROR: test_json_parser (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_json_parser

**Comment by DixonDs** [comment]
I am not quite sure what is going on here - the tests work from Pycharm, but to make them work from "coverage run", I need to modify all imports in the sources like `from call_ai_function import call_ai_function` -> `from scripts.call_ai_function import call_ai_function`

Any hints, please?

**Comment by DixonDs** [comment]
@richbeales it should be fixed now, could you please review it again?
